pkg/storage/ingest/reader.go

```diff
@@ -407,13 +408,18 @@ func (r *PartitionReader) consumeFetches(ctx context.Context, fetches kgo.Fetche
 		MaxBackoff: 2 * time.Second,
 		MaxRetries: 0, // retry forever
 	})
-
 	for boff.Ongoing() {
+		// If the PartitionReader is stopping and the ctx was cancelled, we don't want to interrupt the in-flight
+		// processing midway. Instead, we let it finish, assuming it'll succeed.
+		// If the processing fails while stopping, we log the error and let the backoff stop and bail out.
+		// There is an edge-case when the processing gets stuck and doesn't let the stopping process. In such a case,
+		// we expect the infrastructure (e.g. k8s) to eventually kill the process.
```

Comment:
the log message is a bit deceiving now - it says "will retry," but it won't retry if the context was cancelled. Can we make the message conditional ("not retrying" and "will retry")?

Reply:
I update the log line below, removing the "will retry" part. I think, from the presence of the `num_retries` field, it should be relatively clear, what is happening in the logs.

Reply:
if we're in a retry loop and then shut down the ingester, you can't know whether the last batch was successfully ingested or the backoff was interrupted. I'm still more inclined to make it dead obvious. Another way is to add a line which says "successfully ingested data from kafka."

Reply:
Ah, I see it now. Unfortunately, we cannot tell if the loop will retry, while we are inside the loop. I.e. the context, the `boff` holds to can be cancelled between the calls to `boff.Wait()` and `boff.Ongoing()`. So the loop can bail, even though the log told that it will retry. This will make it even more confusing, IMHO.

Apart from logging every successful attempt, we can explicitly log the `context canceled`, when exit with an error (like the PR did it initially). But Marco had a good point, that it's, perhaps, redundant (see https://github.com/grafana/mimir/pull/8867#discussion_r1698707575).

Reply:
ok, this is more complicated than I anticipated :D let's keep it as-is.